### PR TITLE
[DOC-362] [2024.1] memory division flags, best practices for tablet overhead

### PR DESCRIPTION
**Pull request by mdbridge**
this is for 2024.1

### CODE REVIEW

**Comment by mdbridge** [comment]
https://deploy-preview-22378--infallible-bardeen-164bc9.netlify.app/preview/develop/best-practices-ysql/#minimize-the-number-of-tablets-you-need
https://deploy-preview-22378--infallible-bardeen-164bc9.netlify.app/preview/reference/configuration/yb-tserver/#memory-division-flags

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 1]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
+| TServer %  | 45% | 48% | 57% | 60% |
+| master %   | 20% | 15% | 10% | 10% |
+| Postgres % | 25% | 27% | 28% | 27% |
+| other %    | 10% | 10% |  5% |  3% |
+
+To read this table, take your node's available memory in GiB and find the column who's heading interval it falls under.  For example, a 7 GiB node would fall in the (4,8] interval because it is bigger than 4 and less than or equal to 8.  The defaults for [`--default_memory_limit_to_ram_ratio`](#default-memory-limit-to-ram-ratio) on this node will thus be `0.48` for TServers and `0.15` for masters. The Postgres and other percentages are not set via gflag today but rather consist of whatever memory is left after TServer and master take their cut.
+
+For comparison, when `--use_memory_defaults_optimized_for_ysql` is `false`, the split is TServer 85%, master 10%, Postgres 0%, and other 5%.
+
+The defaults for the within TServer partitioning flags when `--use_memory_defaults_optimized_for_ysql` is true do not depend on the node size and are:
```
Comment (aishwarya24) [inline docs/content/preview/reference/configuration/yb-tserver.md:249]:
```suggestion
The defaults for the TServer partitioning flags when `--use_memory_defaults_optimized_for_ysql` is true do not depend on the node size, and are described in the following table:
```

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 2]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
+
+
+
+### Flags controlling the split of memory among processes
+
+Note that in general these flags will have different values for TServer and master processes.
+
+##### --memory_limit_hard_bytes
+
+Maximum amount of memory this process should use in bytes, its hard memory limit.  A value of `0` specifies to instead use a percentage of the total system memory; see [`--default_memory_limit_to_ram_ratio`](#default-memory-limit-to-ram-ratio) for the percentage used.  A value of `-1` disables all memory limiting.
```
Comment (aishwarya24) [inline docs/content/preview/reference/configuration/yb-tserver.md:268]:
```suggestion
Maximum amount of memory this process should use in bytes, that is, its hard memory limit. A value of `0` specifies to instead use a percentage of the total system memory; see [`--default_memory_limit_to_ram_ratio`](#default-memory-limit-to-ram-ratio) for the percentage used. A value of `-1` disables all memory limiting.
```

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 3]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
+The percentage of available RAM to use for this process if [`--memory_limit_hard_bytes`](#memory-limit-hard-bytes) is `0`.  The special value `-1000` means to instead use the default value for this flag.  Available RAM excludes memory reserved by the kernel.
+
+Default: `0.85` unless [`--use_memory_defaults_optimized_for_ysql`](#use-memory-defaults-optimized-for-ysql) is true.
+
+
+### Flags controlling the split of memory within a TServer
+
+##### --db_block_cache_size_bytes
+
+Size of the shared RocksDB block cache (in bytes).  A value of `-1` specifies to instead use a percentage of this processes' hard memory limit; see [`--db_block_cache_size_percentage`](#db_block_cache_size_percentage) for the percentage used.  A value of `-2` disables the block cache.
```
Comment (aishwarya24) [inline docs/content/preview/reference/configuration/yb-tserver.md:283]:
```suggestion
Size of the shared RocksDB block cache (in bytes). A value of `-1` specifies to instead use a percentage of this processes' hard memory limit; see [`--db_block_cache_size_percentage`](#db_block_cache_size_percentage) for the percentage used. A value of `-2` disables the block cache.
```

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 4]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
+
+Default: `-1`
+
+##### --db_block_cache_size_percentage
+
+Percentage of our hard memory limit to use for the shared RocksDB block cache, if [`--db_block_cache_size_bytes`](#db-block-cache-size-bytes) is that `-1`.  The special value `-1000` means to instead use the default value for this flag.  The special value `-3` means to use a older default that does not take the amount of RAM into account.
```
Comment (aishwarya24) [inline docs/content/preview/reference/configuration/yb-tserver.md:289]:
```suggestion
Percentage of Yugabyte's hard memory limit to use for the shared RocksDB block cache, if [`--db_block_cache_size_bytes`](#db-block-cache-size-bytes) is `-1`.  The special value `-1000` means to instead use the default value for this flag.  The special value `-3` means to use an older default that does not take the amount of RAM into account.
```

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 5]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
+
+Default: `50` unless [`--use_memory_defaults_optimized_for_ysql`](#use-memory-defaults-optimized-for-ysql) is true.
+
+##### --tablet_overhead_size_percentage
+
+Percentage of our hard memory limit to use for tablet-related overheads. A value of `0` means no limit.  Must be between `0` and `100` inclusive. Exception: `-1000` specifies to instead use the default value for this flag.
```
Comment (aishwarya24) [inline docs/content/preview/reference/configuration/yb-tserver.md:295]:
```suggestion
Percentage of Yugabyte's hard memory limit to use for tablet-related overheads. A value of `0` means no limit. Must be between `0` and `100` inclusive. Exception: `-1000` specifies to instead use the default value for this flag.
```

**Review by aishwarya24 — COMMENTED** [review]
Left some suggestions.

**Comment by mdbridge** [comment]
ok, this is ready to review;   I will add a section on master gflags similar to the TServer one once I know if you like the current TServer draft.

File: docs/content/preview/develop/best-practices-ysql.md [hunk 6]
```diff
@@ -258,30 +258,34 @@ For consistent latency or performance, it is recommended to size columns in the
...
+- Use [colocation](../../architecture/docdb-sharding/colocated-tables/) to group small tables into 1 tablet.
+- Reduce number of tablets-per-table using [`--ysql_num_shards_per_tserver`](../../reference/configuration/yb-tserver/#yb-num-shards-per-tserver) flag.
+- Use [`SPLIT INTO`](../../api/ysql/the-sql-language/statements/ddl_create_table/#split-into) clause when creating a table.
+- Start with few tablets and use [automatic tablet splitting](../../architecture/docdb-sharding/tablet-splitting/).
+
+Note that multiple tablets can allow work to proceed in parallel so you may not want every table to have only one tablet.
+
+## Allowing for tablet replica overheads
+
+For a universe with RF3, 1000 tablets imply 3000 tablet replicas. If the universe has three nodes, then each node has on average 1000 tablet replicas. A six node universe would have on average 500 tablet replicas per-node and so on.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/develop/best-practices-ysql.md:278]: fixed, including a reference to RF3 you haven't seen yet

File: docs/content/preview/develop/best-practices-ysql.md [hunk 7]
```diff
@@ -258,30 +258,34 @@ For consistent latency or performance, it is recommended to size columns in the
...
 
-- Use [colocation](../../architecture/docdb-sharding/colocated-tables/) to group small tables into 1 tablet.
-- Reduce number of tablets-per-table using [`--ysql_num_shards_per_tserver`](../../reference/configuration/yb-tserver/#yb-num-shards-per-tserver) flag.
-- Use [`SPLIT INTO`](../../api/ysql/the-sql-language/statements/ddl_create_table/#split-into) clause when creating a table.
-- Start with few tablets and use [automatic tablet splitting](../../architecture/docdb-sharding/tablet-splitting/).
+Accordingly, we recommend you instead use the new [`--use_memory_defaults_optimized_for_ysql`](../../reference/configuration/yb-tserver/#use-memory-defaults-optimized-for-ysql) flag, which gives you good memory division settings for using YSQL optimized for your node's size.  Consult the table given there to see how big of a node you will need based on how many tablet replicas per server you want supported.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/develop/best-practices-ysql.md:288]: made first fix; you misunderstood the second part.  I changed it to:
Consult the table there showing node RAM versus maximum tablet replicas
to see how big of a node you will need based on how many tablet replicas
per server you want supported.

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 8]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
 Default: `false`
 
+
+## Memory division flags
+
+These flags are used to determine how the RAM of a node is split between the TServer and other processes, including Postgres and a master process if present, as well as how to split memory inside of a TServer between various internal components like the RocksDB block cache.
+
+{{< warning title="Warning" >}}
+
+Be careful to ensure you do not oversubscribe memory when changing these flags: make sure the amount of memory reserved for TServer and master if present leaves enough memory on the node for Postgres and any needed kernel and other processes like monitoring agents.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md]: fixed

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 9]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
+
+{{< /warning >}}
+
+
+### Flags controlling the defaults for the other memory division flags
+
+The memory division flags have multiple sets of defaults; which set of defaults is in force depends on these flags.  Note also that these flags have different defaults for master processes.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md]: I found my original version awkward so changed it to:
Note that these defaults can differ between TServer and master.

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 10]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
+
+##### --use_memory_defaults_optimized_for_ysql
+
+If true, the defaults for the memory division settings take into account the amount of RAM and cores available and are optimized for using YSQL.  If false, the recommended defaults will be the old defaults, which are more suitable for YCQL but do not take into account the amount of RAM and cores available.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md]: not fixing; two spaces after periods is standard for monospace fonts like markdown.
verified that it looks fine in the actual HTML output

File: docs/content/preview/reference/configuration/yb-tserver.md (hunk 1 above)
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md:249]: a bit wordy, but change made

File: docs/content/preview/reference/configuration/yb-tserver.md (hunk 2 above)
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md:283]: text added

File: docs/content/preview/reference/configuration/yb-tserver.md (hunk 3 above)
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md:298]: not fixing

File: docs/content/preview/reference/configuration/yb-tserver.md (hunk 4 above)
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md]: changed Yugabyte's to process';  made other fixes

File: docs/content/preview/reference/configuration/yb-tserver.md (hunk 5 above)
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md:310]: ditto

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 11]
```diff
@@ -209,6 +209,96 @@ Stop attempting to log to disk if the disk is full.
...
+If these defaults are used for both TServer and master then a node's available memory is partitioned as follows:
+
+| node RAM GiB: | (0,4] | (4,8] | (8,16] | (16,infinity) |
+| :--- | :--- | :--- | :--- | :--- |
+| TServer %  | 45% | 48% | 57% | 60% |
+| master %   | 20% | 15% | 10% | 10% |
+| Postgres % | 25% | 27% | 28% | 27% |
+| other %    | 10% | 10% |  5% |  3% |
+
+To read this table, take your node's available memory in GiB and find the column who's heading interval it falls under.  For example, a 7 GiB node would fall in the (4,8] interval because it is bigger than 4 and less than or equal to 8.  The defaults for [`--default_memory_limit_to_ram_ratio`](#default-memory-limit-to-ram-ratio) on this node will thus be `0.48` for TServers and `0.15` for masters. The Postgres and other percentages are not set via gflag today but rather consist of whatever memory is left after TServer and master take their cut.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md]: fixed the flag part.  Left the interval alone.  This is a math notation: (x,y]  means the half-open interval starting just after x going up to *and including* y.  I give an example to help clarify this.

if you think this is too confusing, please push back

**Comment by mdbridge** [comment]
preview link for master changes: https://deploy-preview-22378--infallible-bardeen-164bc9.netlify.app/preview/reference/configuration/yb-master/#memory-division-flags

**Review by aishwarya24 — APPROVED** [review]
LGTM with minor suggestions.

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 12]
```diff
@@ -209,6 +209,111 @@ Stop attempting to log to disk if the disk is full.
 
 Default: `false`
 
+
+## Memory division flags
+
+These flags are used to determine how the RAM of a node is split between the TServer and other processes, including Postgres and a master process if present, as well as how to split memory inside of a TServer between various internal components like the RocksDB block cache.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md:215]: fixed, including on master page

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 13]
```diff
@@ -209,6 +209,111 @@ Stop attempting to log to disk if the disk is full.
...
+
+{{< warning title="Warning" >}}
+
+Ensure you do not _oversubscribe memory_ when changing these flags: make sure the amount of memory reserved for TServer and master if present leaves enough memory on the node for Postgres and any needed other processes like monitoring agents plus the memory needed by the kernel.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md:219]: fixed

File: docs/content/preview/reference/configuration/yb-tserver.md [hunk 14]
```diff
@@ -209,6 +209,111 @@ Stop attempting to log to disk if the disk is full.
...
+If these defaults are used for both TServer and master then a node's available memory is partitioned as follows:
+
+| node RAM GiB (_M_): | _M_ &nbsp;&le;&nbsp; 4 | 4 < _M_ &nbsp;&le;&nbsp; 8 | 8 < _M_ &nbsp;&le;&nbsp; 16 | 16 < _M_ |
+| :--- | ---: | ---: | ---: | ---: |
+| TServer %  | 45% | 48% | 57% | 60% |
+| master %   | 20% | 15% | 10% | 10% |
+| Postgres % | 25% | 27% | 28% | 27% |
+| other %    | 10% | 10% |  5% |  3% |
+
+To read this table, take your node's available memory in GiB, call it _M_, and find the column who's heading condition _M_ meets.  For example, a node with 7 GiB of available memory would fall under the column labeled "4 < _M_ &le; 8" because 4 < 7 &le; 8.  The defaults for [`--default_memory_limit_to_ram_ratio`](#default-memory-limit-to-ram-ratio) on this node will thus be `0.48` for TServers and `0.15` for masters. The Postgres and other percentages are not set via gflag currently but rather consist of whatever memory is left after TServer and master take their cut.  There is currently no distinction between Postgres and other memory except on YBM where a cgroup is used to limit the Postgres memory.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-tserver.md:245]: fixed

File: docs/content/preview/reference/configuration/yb-master.md [hunk 15]
```diff
@@ -251,6 +251,111 @@ Log messages at, or above, this level are copied to `stderr` in addition to log
...
+For comparison, when `--use_memory_defaults_optimized_for_ysql` is `false`, the split is TServer 85%, master 10%, Postgres 0%, and other 5%.
+
+The defaults for the master process partitioning flags when `--use_memory_defaults_optimized_for_ysql` is `true` do not depend on the node size, and are described in the following table:
+
+| flag | default |
+| :--- | :--- |
+| --db_block_cache_size_percentage | 25 |
+| --tablet_overhead_size_percentage | 0 |
+
+Today these are the same as the defaults when `--use_memory_defaults_optimized_for_ysql` is `false`, but may change in future releases.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-master.md:298]: fixed

File: docs/content/preview/reference/configuration/yb-master.md [hunk 16]
```diff
@@ -251,6 +251,111 @@ Log messages at, or above, this level are copied to `stderr` in addition to log
...
+
+Given the amount of RAM devoted to per tablet overhead, it is possible to compute the maximum number of tablet replicas (see [allowing for tablet replica overheads](../../../develop/best-practices-ysql#allowing-for-tablet-replica-overheads)); here are some sample values for selected node sizes using `--use_memory_defaults_optimized_for_ysql`:
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-master.md:300]: your version doesn't read as well as mine to my eye but changed it anyways

File: docs/content/preview/reference/configuration/yb-master.md [hunk 17]
```diff
@@ -251,6 +251,111 @@ Log messages at, or above, this level are copied to `stderr` in addition to log
...
+The percentage of available RAM to use for this process if [`--memory_limit_hard_bytes`](#memory-limit-hard-bytes) is `0`.  The special value `-1000` means to instead use the default value for this flag.  Available RAM excludes memory reserved by the kernel.
+
+Default: `0.10` unless [`--use_memory_defaults_optimized_for_ysql`](#use-memory-defaults-optimized-for-ysql) is true.
+
+
+### Flags controlling the split of memory within a master process
+
+##### --db_block_cache_size_bytes
+
+Size of the shared RocksDB block cache (in bytes).  A value of `-1` specifies to instead use a percentage of this processes' hard memory limit; see [`--db_block_cache_size_percentage`](#db_block_cache_size_percentage) for the percentage used.  A value of `-2` disables the block cache.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-master.md:340]: fixed

File: docs/content/preview/reference/configuration/yb-master.md [hunk 18]
```diff
@@ -251,6 +251,111 @@ Log messages at, or above, this level are copied to `stderr` in addition to log
 
 Default: `2`
 
+
+## Memory division flags
+
+These flags are used to determine how the RAM of a node is split between the master process and other processes, including Postgres and a TServer process if present, as well as how to split memory inside of a master process between various internal components like the RocksDB block cache.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-master.md:257]: fixed

File: docs/content/preview/reference/configuration/yb-master.md [hunk 19]
```diff
@@ -251,6 +251,111 @@ Log messages at, or above, this level are copied to `stderr` in addition to log
...
+
+{{< warning title="Warning" >}}
+
+Ensure you do not _oversubscribe memory_ when changing these flags: make sure the amount of memory reserved for the master process and TServer if present leaves enough memory on the node for Postgres and any needed other processes like monitoring agents plus the memory needed by the kernel.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-master.md:261]: fixed

File: docs/content/preview/reference/configuration/yb-master.md [hunk 20]
```diff
@@ -251,6 +251,111 @@ Log messages at, or above, this level are copied to `stderr` in addition to log
...
+If these defaults are used for both TServer and master then a node's available memory is partitioned as follows:
+
+| node RAM GiB (_M_): | _M_ &nbsp;&le;&nbsp; 4 | 4 < _M_ &nbsp;&le;&nbsp; 8 | 8 < _M_ &nbsp;&le;&nbsp; 16 | 16 < _M_ |
+| :--- | ---: | ---: | ---: | ---: |
+| TServer %  | 45% | 48% | 57% | 60% |
+| master %   | 20% | 15% | 10% | 10% |
+| Postgres % | 25% | 27% | 28% | 27% |
+| other %    | 10% | 10% |  5% |  3% |
+
+To read this table, take your node's available memory in GiB, call it _M_, and find the column who's heading condition _M_ meets.  For example, a node with 7 GiB of available memory would fall under the column labeled "4 < _M_ &le; 8" because 4 < 7 &le; 8.  The defaults for [`--default_memory_limit_to_ram_ratio`](#default-memory-limit-to-ram-ratio) on this node will thus be `0.48` for TServers and `0.15` for masters. The Postgres and other percentages are not set via gflag currently but rather consist of whatever memory is left after TServer and master take their cut.  There is currently no distinction between Postgres and other memory except on YBM where a cgroup is used to limit the Postgres memory.
```
[the remark this answers is not in the document]
Comment (mdbridge) [reply docs/content/preview/reference/configuration/yb-master.md:287]: fixed

**Review by druzac — APPROVED** [review]
